galsim/photon_array.py

```diff
@@ -0,0 +1,86 @@
+# conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions, and the disclaimer given in the accompanying LICENSE
+#    file.
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions, and the disclaimer given in the documentation
+#    and/or other materials provided with the distribution.
+#
+
+import numpy as np
+from ._galsim import PhotonArray
+import galsim
+
+class FRatioAngles(object):
+    """A surface-layer operator that assigns photon directions based on the f/ratio and
+    obscuration.
+
+    Assigns arrival directions at the focal plane for photons, drawing from a uniform
+    brightness distribution between the obscuration angle and the angle of the FOV defined
+    by the f/ratio of the telescope.  The angles are expressed in terms of slopes dx/dz
+    and dy/dz.
+
+    @param fratio           The f-ratio of the telescope (e.g. 1.2 for LSST)
+    @param obscuration      Linear dimension of central obscuration as fraction of aperture
+                            linear dimension. [0., 1.).  [default: 0.0]
+    @param rng              A random number generator to use or None, in which case an rng
+                            will be automatically constructed for you. [default: None]
+    """
+    def __init__(self, fratio, obscuration=0.0, rng=None):
+
+        if fratio < 0:
+            raise ValueError("The f-ratio must be positive.")
+        if obscuration < 0 or obscuration >= 1:
+            raise ValueError("The obscuration fraction must be between 0 and 1.")
+        if rng is None:
+            ud = galsim.UniformDeviate()
+        else:
+            ud = galsim.UniformDeviate(rng)
+
+        self.fratio = fratio
+        self.obscuration = obscuration
+        self.ud = ud
+
+
+    def applyTo(self, photon_array):
+        """Assign directions to the photons in photon_array."""
+
+        dxdz = photon_array.getDXDZArray()
+        dydz = photon_array.getDYDZArray()
+        n_photons = len(dxdz)
+
+        fov_angle = np.arctan(0.5 / self.fratio)  # radians
+        obscuration_angle = self.obscuration * fov_angle
+
+        # Generate azimuthal angles for the photons
+        # Set up a loop to fill the array of azimuth angles for now
+        # (The array is initialized below but there's no particular need to do this.)
+        phi = np.zeros(n_photons)
+
+        for i in np.arange(n_photons):
+            phi[i] = self.ud() * 2 * np.pi
+
+        # Generate inclination angles for the photons, which are uniform in sin(theta) between
+        # the sine of the obscuration angle and the sine of the FOV radius
+        sintheta = np.zeros(n_photons)
+
+        for i in np.arange(n_photons):
+            sintheta[i] = np.sin(obscuration_angle) + (np.sin(fov_angle) - \
+                          np.sin(obscuration_angle))*self.ud()
```

Comment:
Likewise, this would be more efficient as
```
sinobs = np.sin(obscuration_angle)
sinfov = np.sin(fov_angle)
sintheta = sinobs + (sinfov - sinobs) * self.ud.generate(n_photons)
```